### PR TITLE
Bump minimum kubeadm supported version to v1.13

### LINKED ISSUE #1297: Bump minimum supported version to v1.13 at HEAD
xref: https://github.com/kubernetes/kubeadm/issues/1260
this should be simpler than it is but meanwhile...

let's:
 - bump the minimum version to v1.13 at HEAD
 - remove v1.12-specific code
 - remove v1.10, v1.11, v1.12 records for matching etcd versions

@yagonobre or @ereslibre can you help with this?

### PR DESCRIPTION
**Pull request by yagonobre**
**What type of PR is this?**

/kind cleanup
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubeadm/issues/1297

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/apis/kubeadm/v1alpha3/conversion_test.go
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha3/conversion_test.go]: can we use some of the helpers @bart0sh added to the version package?
https://github.com/kubernetes/kubernetes/commit/af07603178b6fca272106a941fdac87c957f7766#diff-533e2521f45a66cc7c4c2591b07ae7f8

you can search for the current usage that we have in kubeadm.
the idea is to use `MinimumControlPlaneVersion` + these methods.

i guess this comment applies to the rest of the changes too.

File: cmd/kubeadm/app/constants/constants.go
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants.go]: this is basically the support skew - current version minus 2.
having said that i think we should keep 1.12 even if it points to the same version as 1.13.

File: cmd/kubeadm/app/constants/constants_test.go
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants_test.go]: we can technically this etcd test better.
possibly in another PR if you'd prefer.

what can be done is to add test cases by iterating the `SupportedEtcdVersion` map for valid cases and using @bart0sh 's `With...` methods. just an idea.

this way we don't have to bump this test on each minor release.

**Review by neolit123 — APPROVED** [review]
thanks @yagonobre mostly LGTM.
added a couple of comments.

/assign @fabriziopandini @rosti

**Review by rosti — COMMENTED** [review]
Much like @neolit123, I'd like to see version strings changed to `Minimum*Version` and `CurrentKubernetesVersion` rather than have lots of place where we bump strings.

File: cmd/kubeadm/app/cmd/phases/util_test.go
Comment (bart0sh) [inline cmd/kubeadm/app/cmd/phases/util_test.go]: should we use constants.CurrentKubernetesVersion here?

File: cmd/kubeadm/app/componentconfigs/config_test.go
Comment (bart0sh) [inline cmd/kubeadm/app/componentconfigs/config_test.go]: and here

**Comment by yagonobre** [comment]
@neolit123, @rosti and @bart0sh Thanks for the review, I'll take a look as soon as possible.

File: cmd/kubeadm/app/cmd/phases/util_test.go
Comment (yagonobre) [reply cmd/kubeadm/app/cmd/phases/util_test.go]: I guess that I no need to change this test

File: cmd/kubeadm/app/constants/constants_test.go
Comment (yagonobre) [reply cmd/kubeadm/app/constants/constants_test.go]: I'll do that in another pr, probably in the next cycle

**Comment by yagonobre** [comment]
@neolit123 @rosti @bart0sh Updated, Should I bump the current version to 1.14?

**Comment by yagonobre** [comment]
I'm trying to refactor `TestGetAvailableUpgrades` but it's a little complicated I'll try again tomorrow.

File: cmd/kubeadm/app/cmd/config_test.go
Comment (bart0sh) [inline cmd/kubeadm/app/cmd/config_test.go]: String() is redundant here

**Comment by neolit123** [comment]
> Should I bump the current version to 1.14?

yes, we can do this now in `master`.

**Review by timothysc — APPROVED** [review]
/approve 

I'll let others lgtm when ready.

**Comment by neolit123** [comment]
@yagonobre 
please let me know if you don't have the time for this PR and i will create a new one.
`master` is already tracking 1.15.

File: cmd/kubeadm/app/phases/upgrade/policy_test.go
Comment (yagonobre) [inline cmd/kubeadm/app/phases/upgrade/policy_test.go]: I need to revert some of this changes because bump `CurrentKubernetesVersion` to `1.14.0` broke this test. We probably should use something like `constants.CurrentKubernetesVersion.WitchMinorMinus(1).WithPatch(3).String()` in  future.

**Comment by yagonobre** [comment]
Updated! I'll let test refactor to next cycle (sorry for this).

File: cmd/kubeadm/app/constants/constants.go
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants.go]: this now has to be `version.MustParseSemantic("v1.13.0")`
same for `MinimumControlPlaneVersion`

Comment (yagonobre) [reply cmd/kubeadm/app/constants/constants.go]: `CurrentKubernetesVersion` should be `1.15.0`?

Comment (neolit123) [reply cmd/kubeadm/app/constants/constants.go]: not until we stop fast-forwarding the master branch into release-1.14.

current = 1.14
min-kubelet = current -1
min-control-plane = current - 1

Comment (yagonobre) [reply cmd/kubeadm/app/constants/constants.go]: done!

File: cmd/kubeadm/app/phases/upgrade/compute_test.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/upgrade/compute_test.go]: yeah, looks like we still have to bump a lot of versions, while constants constants should be used instead.
https://github.com/kubernetes/kubeadm/issues/1260

**Review by neolit123 — APPROVED** [review]
thanks @yagonobre 

yeah, looks like we still have to bump a lot of versions, while constants should be used instead.
https://github.com/kubernetes/kubeadm/issues/1260

/lgtm
/approve
/hold

@fabriziopandini @rosti 
please feel free to `/hold cancel`.

**Comment by yagonobre** [comment]
I updated `TestGetAvailableUpgrades` to use constants  on  tests cases  that only depends  on current and minimum version, it's decreases a lot the total of place that we need to edit. I'll let the rest to another interaction.

**Review by rosti — APPROVED** [review]
Thanks @yagonobre !
/lgtm

@fabriziopandini @neolit123 let's get this in as fast as possible before CF.

**Comment by fabriziopandini** [comment]
Thanks @yagonobre 
This is step in the right direction for simplifying version bump
/approve
/lgtm
/hold cancel

**Pull request merged**
